Sources/Swift/HangTracker.swift

```diff
@@ -0,0 +1,302 @@
+                // We already applied the scope. We use an empty scope to avoid overwriting exising
+                // fields on the event.
+                SentrySDK.capture(event: event, scope: Scope())
+            } else {
+                // Fatal App Hang
+                // We can't differ if the watchdog or the user terminated the app, because when the main
+                // thread is blocked we don't receive the applicationWillTerminate notification. Further
+                // investigations are required to validate if we somehow can differ between watchdog or
+                // user terminations; see https://github.com/getsentry/sentry-cocoa/issues/4845.
+                guard let exceptions = event.exceptions, let exception = exceptions.first, exceptions.count == 1 else {
+                    SentrySDKLog.warning("The stored app hang event is expected to have exactly one exception, so we don't capture it.")
+                    return
+                }
+                
+                SentryLevelBridge.setBreadcrumbLevelOn(event, level: SentryLevel.fatal.rawValue)
+                event.exceptions?.first?.mechanism?.handled = false
+                let fatalExceptionType = SentryAppHangTypeMapper.getFatalExceptionType(nonFatalErrorType: exception.type)
+                event.exceptions?.first?.type = fatalExceptionType
+                
+                var mechanismData = exception.mechanism?.data
+                let durationInfo = mechanismData?[Self.SentryANRMechanismDataAppHangDuration] as? String ?? "over \(minHangTime) seconds"
+                mechanismData?.removeValue(forKey: Self.SentryANRMechanismDataAppHangDuration)
+                event.exceptions?.first?.value = "The user or the OS watchdog terminated your app while it blocked the main thread for \(durationInfo)"
+                event.exceptions?.first?.mechanism?.data = mechanismData
+                SentryDependencyContainerSwiftHelper.captureFatalAppHang(event)
+                
+            }
+        }
+    }
+    
+    // MARK: Background queue
+    
+    private var blockingDuration: TimeInterval?
+    private var threads: [SentryThread]?
+    
+    private func waitForHang(semaphore: DispatchSemaphore, started: TimeInterval, isStarting: Bool) {
+        dispatchPrecondition(condition: .onQueue(queue))
+        
+        let timeout = DispatchTime.now() + DispatchTimeInterval.milliseconds(Int((expectedFrameDuration + thresholdForFrameStacktrace) * 1_000))
+        let result = semaphore.wait(timeout: timeout)
+        switch result {
+        case .timedOut:
+            semaphore.signal()
+            print("[HANG] Timeout, hang detected")
+            continueHang(started: started, isStarting: isStarting)
+            waitForHang(semaphore: semaphore, started: started, isStarting: false)
+        case .success:
+            break
+        }
+    }
+
+    private func continueHang(started: TimeInterval, isStarting: Bool) {
```

Comment:
How does this work if the app is backgrounded?

Reply:
Gotcha, I was more thinking about whether the `lastFrameTime` value could get out of sync at all and report a false positive since that seems like it might have happened with the current implementation, but I'm guessing you're right that the runloop is allowed to finish its iteration.